### PR TITLE
[TASK-7159] [TASK-7192] [TASK-7193] feat: send money with peanut-wallet

### PR DESCRIPTION
**Pull request by jjramirezn**
## Main changes

- Refactor in wallet context to correctly handle wallet state

## Other changes
- Added some packages previously removed because they are peerDependencies
- Removed duplicated element (possibly from conflict resolving)
- Renamed areTokenAddressesEqual to areAdressesEqual (not only token addresses are compared)

## QA info
Right now it should be possible to generate a request link, fulfill it and then generate a claim link with that balance.
IMPORTANT: right now the token must be manually selected, in a (near)future PR we will hide the token selector. Select USDC in arbitrum as token.

## Release Notes

- **New Features**
	- Introduced a new logging utility for better log readability.
	- Enhanced wallet management functionality, including improved wallet connection checks.

- **Improvements**
	- Updated address comparison logic across multiple components for consistency and robustness.
	- Enhanced error handling and messaging in various components to provide clearer feedback to users.
	- Improved API interaction and validation checks for user accounts and liquidation addresses.

These updates aim to improve user experience and streamline interactions within the application.

### CODE REVIEW

File: src/components/Claim/Link/Initial.view.tsx [hunk 1]
```diff
@@ -7,9 +7,8 @@ import Icon from '@/components/Global/Icon'
 import useClaimLink from '../useClaimLink'
 import * as context from '@/context'
 import * as consts from '@/constants'
-import { supportedPeanutChains } from '@/constants'
 import {
-    areTokenAddressesEqual,
+    areAddressesEqual,
```
Comment (Hugo0) [inline src/components/Claim/Link/Initial.view.tsx]: nice

File: src/components/Claim/Link/Initial.view.tsx [hunk 2]
```diff
@@ -516,44 +515,6 @@ export const InitialClaimLinkView = ({
                             </div>
                         )}
 
-                        <div className="flex w-full flex-row items-center justify-between px-2 text-h8 text-gray-1">
```
Comment (Hugo0) [inline src/components/Claim/Link/Initial.view.tsx:519]: Q: why are we removing this?

File: src/components/Create/Create.utils.ts [hunk 3]
```diff
@@ -51,19 +51,17 @@ function toLowerCaseKeys(obj: any): any {
...
         tokenDecimals =
             userBalances.find(
-                (balance) => balance.chainId == chainId && areTokenAddressesEqual(balance.address, tokenAddress)
+                (balance) => balance.chainId == chainId && areAddressesEqual(balance.address, tokenAddress)
             )?.decimals ?? 18
     } else {
         tokenDecimals =
             peanutTokenDetails
                 .find((detail) => detail.chainId.toString() == chainId)
-                ?.tokens.find((token) => areTokenAddressesEqual(token.address, tokenAddress))?.decimals ?? 18
```
Comment (Hugo0) [inline src/components/Create/Create.utils.ts:66]: > .find((detail) => detail.chainId.toString() == chainId)
Nit Q: shouldn't this be triple equality?

File: src/components/Create/Create.utils.ts [hunk 4]
```diff
@@ -51,19 +51,17 @@ function toLowerCaseKeys(obj: any): any {
 export const getTokenDetails = (tokenAddress: string, chainId: string, userBalances: IUserBalance[]) => {
     let tokenDecimals: number = 18
     if (
-        userBalances.some(
-            (balance) => areTokenAddressesEqual(balance.address, tokenAddress) && balance.chainId == chainId
-        )
+        userBalances.some((balance) => areAddressesEqual(balance.address, tokenAddress) && balance.chainId == chainId)
```
Comment (Hugo0) [inline src/components/Create/Create.utils.ts]: nit: triple equality?

File: src/utils/general.utils.ts [hunk 5]
```diff
@@ -368,7 +368,7 @@ export const isTestnetChain = (chainId: string) => {
     return isTestnet
 }
 
-export const areTokenAddressesEqual = (address1: string, address2: string): boolean => {
+export const areAddressesEqual = (address1: string, address2: string): boolean => {
```
Comment (Hugo0) [inline src/utils/general.utils.ts]: thought: "address" could be confusing (street address, bank address etc). Not sure what name would be better though

**Comment by Hugo0** [comment]
@coderabbitai review

File: src/context/walletContext/walletContext.tsx [hunk 6]
```diff
@@ -28,18 +26,12 @@ interface WalletContextType {
     walletColor: string
 }
 
-function isPeanut(wallet: interfaces.IWallet | undefined) {
+function isPeanut(wallet: interfaces.IDBWallet | undefined) {
```
Comment (Hugo0) [inline src/context/walletContext/walletContext.tsx:29]: thought: wdyt of maintaining an interface in FE and also maintaining the backend implementation? Seems uneccesarily duplicate

File: src/context/walletContext/walletContext.tsx [hunk 7]
```diff
@@ -28,18 +26,12 @@ interface WalletContextType {
...
     return wallet?.walletProviderType === interfaces.WalletProviderType.PEANUT
 }
```
Comment (Hugo0) [inline src/context/walletContext/walletContext.tsx:31]: nit: function name seems quite short - I'd prefer overly long but descriptive names over too short generic ones that might cause confusion down the line

File: src/context/walletContext/walletContext.tsx [hunk 8]
```diff
@@ -1,16 +1,14 @@
 'use client'
```
Comment (Hugo0) [inline src/context/walletContext/walletContext.tsx:1]: thought: we expect this context to be available everywhere right? I find the organization of src > context > walletContext a bit confusing. Especially when we have components > Global too. It's  a bit hard to navigate so many folders, I think we might want to dedicate some time to restructuring this

File: src/context/walletContext/walletContext.tsx [hunk 9]
```diff
@@ -51,166 +43,137 @@ export const WalletProvider = ({ children }: { children: ReactNode }) => {
...
 
-    /**
-     * Update selected wallet with connection status when kernel client is ready or wagmi is connected
-     */
+    // Set initial selected address
     useEffect(() => {
-        if (selectedWallet) {
-            updateSelectedWalletWithConnectionStatus(selectedWallet)
+        if (!selectedAddress && wallets.length > 0) {
+            const initialWallet = wallets.find(isPeanut) ?? wallets[0]
```
Comment (Hugo0) [inline src/context/walletContext/walletContext.tsx:131]: Q: its smart to use peanut wallet as the default wallet, i like it. Question, do we respect users preferences? Ie should we store their last used wallet and let them use that instead on the next session?

File: src/context/walletContext/walletContext.tsx [hunk 10]
```diff
@@ -51,166 +43,137 @@ export const WalletProvider = ({ children }: { children: ReactNode }) => {
...
-
-    const removeBYOW = async () => {
-        // TODO: when successful API call, do NOT refetch all wallets (bad UX), but
-        // go on the wallet list and remove wallet
-        // TODO: if wallet connected to provider, disconnect wallet from provider
-        // then
-        // TODO: if active, default active -> PW if logged in, otherwise
-        // if another wallet is connected to the provider, make that active, otherwise
-        // no wallet active (have to review all props to ensure they are null)
+    }, [wagmiAddress, wallets, user, addAccount])
```
Comment (Hugo0) [inline src/context/walletContext/walletContext.tsx]: Q: doesn't having addAccount in dep array make this circular?

File: src/context/walletContext/walletContext.tsx [hunk 11]
```diff
@@ -51,166 +43,137 @@ export const WalletProvider = ({ children }: { children: ReactNode }) => {
...
+    )
+
+    const createDefaultDBWallet = (address: string): interfaces.IDBWallet => ({
+        walletProviderType: interfaces.WalletProviderType.BYOW,
+        protocolType: interfaces.WalletProtocolType.EVM,
+        address,
+    })
+
+    const { data: wallets = [] } = useQuery<interfaces.IWallet[]>({
         queryKey: ['wallets', user?.accounts, wagmiAddress],
```
Comment (Hugo0) [inline src/context/walletContext/walletContext.tsx:67]: nit: queryKey should containt all dependencies of isWalletConnected

I added a dev dependency to detect this automatically

Comment (Hugo0) [reply src/context/walletContext/walletContext.tsx:67]: nvm the linter doesn't seeem to work

File: src/context/walletContext/walletContext.tsx (hunk 8 above)
Comment (Hugo0) [reply src/context/walletContext/walletContext.tsx:1]: thought 2: I think it was nice to have a high level overview comment explaining the goal of this file / WalletProvider. Sad to see it gone

**Review by Hugo0 — COMMENTED** [review]
Overall lgtm!

**Review by Hugo0 — COMMENTED** [review]
Overall lgtm!

**Review by Hugo0 — APPROVED** [review]

File: package.json [hunk 12]
```diff
@@ -55,6 +55,7 @@
         "lottie-react": "^2.4.0",
         "multicoin-address-validator": "^0.5.22",
         "next": "^14.2.18",
+        "pino-pretty": "^13.0.0",
```
[the remark this answers is not in the document]
Comment (jjramirezn) [reply package.json]: moving

File: src/components/Claim/Link/Initial.view.tsx (hunk 2 above)
Comment (jjramirezn) [reply src/components/Claim/Link/Initial.view.tsx:519]: This was duplicated after merging

File: src/context/walletContext/walletContext.tsx (hunk 8 above)
Comment (jjramirezn) [reply src/context/walletContext/walletContext.tsx:1]: Hmmmm I think the distintion between components and context is useful (even if in the end they are all components). I know that a context will have a state available everywhere and components will be reusable UI pieces. (I do agree that we have many folders and sometimes is hard to navigate, if you ask me to write a path to a component I am working on I wont be able to do it)

File: src/context/walletContext/walletContext.tsx (hunk 6 above)
Comment (jjramirezn) [reply src/context/walletContext/walletContext.tsx:29]: Hmmm, it is duplicate and prone to errors. One way of tackling this is publishing a types package from the backend and use that in the front. That is less error prone but also less flexible. Not sure about using the sdk for this

File: src/context/walletContext/walletContext.tsx (hunk 7 above)
Comment (jjramirezn) [reply src/context/walletContext/walletContext.tsx:31]: I rely a lot in typing a function isPeanut that receives a Wallet object is self describing to me

File: src/context/walletContext/walletContext.tsx (hunk 9 above)
Comment (jjramirezn) [reply src/context/walletContext/walletContext.tsx:131]: Hmmm I think that's a good Idea

File: src/context/walletContext/walletContext.tsx (hunk 10 above)
Comment (jjramirezn) [reply src/context/walletContext/walletContext.tsx]: addAccount should not change, is an util function, it should not be part of this depArray

File: src/utils/general.utils.ts (hunk 5 above)
Comment (jjramirezn) [reply src/utils/general.utils.ts]: renaming to areEvmAddressesEqual

**Pull request merged**